src/hash-table.h

```diff
@@ -235,15 +246,15 @@ void hash_table_iterate(HashTable *hash_table, HashTableIterator *iter);
 int hash_table_iter_has_more(HashTableIterator *iterator);
 
 /**
- * Using a hash table iterator, retrieve the next key.
+ * Using a hash table iterator, retrieve the next @ref KeyValuePair.
  *
  * @param iterator            The hash table iterator.
- * @return                    The next key from the hash table, or
+ * @return                    The next @ref KeyValuePair from the hash table, or
  *                            @ref HASH_TABLE_NULL if there are no more
  *                            keys to iterate over.
  */
 
-HashTableValue hash_table_iter_next(HashTableIterator *iterator);
+KeyValuePair *hash_table_iter_next(HashTableIterator *iterator);
```

Comment:
So I do like this approach. It solves all the problems that we discussed in the previous PR. My main concern is that the KeyValuePair provides a pointer into the internals of the data structure. For example you can do something like:

KeyValuePair *pair = hash_table_iter_next(...);

pair->key = "some new value";

And then suddenly the data structure's internals are messed up.

One approach might be to have hash_table_iter_next return a plain KeyValuePair (not a pointer), copying the key/value that are stored internally. Passing structs on the stack is sometimes frowned on, but this is a fairly small struct (just two pointers) so it's not a big deal really.

Another approach might be to just document it in the doc comment above.